### ISSUE TITLE
Unable to authenticate the request - certificate signed by unknown authority

### ISSUE THREAD

**Issue by jianhang33** [gardener/oidc-webhook-authenticator#94]
**What happened**:
I deployed an instance of oidc-webhook-authenticator in my cluster, and when I make a request to my cluster, I saw a message in the log of oidc-webhook-authenticator

> I0128 00:43:20.035186       1 pathrecorder.go:241] oidc-webhook-authenticator: "/validate-token" satisfied by exact match 

However, in apiserver's log, I saw this

> E0128 00:43:20.037068       1 authentication.go:63] "Unable to authenticate the request" err="invalid bearer token" 

and the output of kubectl is

> error: You must be logged in to the server (Unauthorized)

The token I used to run kubectl is fresh.

I struggled to figure out what certificates to use. Self-signed certificates doesn't seem to work.

**What you expected to happen**:
authentication should pass

**How to reproduce it (as minimally and precisely as possible)**:
I used the root certificate of control plane node (/etc/kubernetes/pki/ca.crt and ca.key) to create a server certificate, and put both ca.crt and server.crt to tls.crt used by the authenticator. For webhook plugin, we used /etc/kubernetes/pki/apiserver-kubelet-client.crt and apiserver-kubelet-client.key.

**Anything else we need to know**:

**Environment**: Linux, Kubernetes

**Comment by vpnachev** [comment]
@jianhang33 can you provide us with detailed step by step instruction how to reproduce this issue?

**Comment by jianhang33** [comment]
Here is the config for webwook plugin config file named gardener-authn-webhook.yaml
```
apiVersion: v1
kind: Config
clusters:
  - name: gardener-authn-service
    cluster:
      insecure-skip-tls-verify: true
      server: https://100.89.145.84:443/validate-token
users:
  - name: api-server
    user:
      client-certificate: /etc/kubernetes/pki/apiserver-kubelet-client.crt  
      client-key: /etc/kubernetes/pki/apiserver-kubelet-client.key          
current-context: webhook-authn
contexts:
- context:
    cluster: gardener-authn-service
    user: api-server
  name: webhook-authn
```

and when the k8s cluster starts, these flags are included
```
--authentication-token-webhook-config-file=/etc/kubernetes/authn/gardener-authn-webhook.yaml
--authentication-token-webhook-cache-ttl=10s
--authentication-token-webhook-version=v1
```

After the k8s cluster is up, I did the following

1. apply config/samples/minikube-webhook-kubeconfig.yaml
2. generate a server certificate from the root certificate/key of the k8s cluster with SAM including 100.89.145.84
3. modify the file config/samples/deployment.yaml to replace the certificate/key in the secret oidc-webhook-authenticator with the server certificate/key created above and the k8s's root certificate
4. remove anything related to dex (as I use my own identity provider) in deployment.yaml
5. deploy deployment.yaml
6. deploy OpenIDConnect object for my identity provider (using jwks)
7. get a token from my identity provider, and run 'kubectl get pods --kubeconfig <kubeconfig.noUserCertificate-key> --token <token>

**Comment by dimityrmirchev** [comment]
Hi, @jianhang33 ! Our local development guide was quite outdated so I filed a PR (https://github.com/gardener/oidc-webhook-authenticator/pull/95) that aims to better reflect the current state of the project.

Do logs from both `kube-apiserver` and `oidc-webhook-authenticator` indicate some errors other than "Unable to authenticate the request" err="invalid bearer token""? If you have the `--authentication-token-webhook-config-file` flag configured then the lack of any other logs in the `kube-apiserver` indicates that the auth requests are successfully being received by the `oidc-webhook`. Is it possible that you misconfigured the `openidconnect` object? Typical misconfigurations include mismatch of issuer ("https://abc.svc" is different from "https://abc.svc/") or wrong audience (`aud` claim from the token should match the `clientID` in the `openidconnect` object).

**Comment by jianhang33** [comment]
Thanks @dimityrmirchev for your response. I didn't see any error messages in oidc-webhook-authenticator's log, and 

> I0128 00:43:20.035186 1 pathrecorder.go:241] oidc-webhook-authenticator: "/validate-token" satisfied by exact match

appears to be the only relevant messages. Is there a way to log more debugging messages in oidc-webhook-authenticator, e.g. log the failure reason when authentication fails? In the meantime, I'll double check the issuer and the audience.

In the log from kube-apiserver, I didn't deploy the webhook for authorization so there is a timeout error message contacting the authorization webhook, but I assume that is independent of the error for authentication

> E0128 00:43:20.037068 1 authentication.go:63] "Unable to authenticate the request" err="invalid bearer token"

**Comment by dimityrmirchev** [comment]
I am not sure that increasing the logging verbosity (controlled by [this](https://github.com/gardener/oidc-webhook-authenticator/blob/a05a395dd7afd451f7a6682aa1200b485eec0024/config/samples/deployment.yaml#L98) flag) will show anything meaningful in this case since the errors during authentication are not logged (see [here](https://github.com/gardener/oidc-webhook-authenticator/blob/a05a395dd7afd451f7a6682aa1200b485eec0024/controllers/authentication/openidconnect_controller.go#L237)). You can try and log the errors there and build and test with the new image if that is something that you want to do.

> In the log from kube-apiserver, I didn't deploy the webhook for authorization so there is a timeout error message contacting the authorization webhook, but I assume that is independent of the error for authentication

The `oidc-webhook-authenticator` also uses authentication and authorization kubeconfigs to check the identity and permissions of the caller so it is making a `SubjectAccessReview` to the `kube-apiserver`. I am not sure if this creates a subsequent request to the `kube-apiserver` authorization webhook and hence the timeouts. I did test the setup described [here](https://github.com/gardener/oidc-webhook-authenticator#local-development) yesterday and did not see any timeouts in the `kube-apiserver` logs.

**Comment by jianhang33** [comment]
You are right on the logging verbosity. I have already increased it to 10. Adding log message to the code is something I can do, but when I tried to compile the code by running 'make all', I got some errors.

```
Successfully built 5c3d6c956105
Successfully tagged tools:latest
docker run --rm --name=oidc-tools -v /home/jianhang/github.com/gardener/oidc-webhook-authenticator:/workspace tools:latest controller-gen object:headerFile="hack/license_boilerplate.go.txt" paths="./..."
panic: interface conversion: ast.Expr is *ast.CompositeLit, not *ast.Ident

goroutine 2610 [running]:
sigs.k8s.io/controller-tools/pkg/loader.(*referenceCollector).Visit(0xc008d1eba0, {0xd6c578?, 0xc00b8ca318})
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:136 +0x105
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6c578?, 0xc00b8ca318?})
	/usr/local/go/src/go/ast/walk.go:51 +0x62
go/ast.walkExprList({0xd69ea0, 0xc008d1eba0}, {0xc007167f30?, 0x1, 0xc00c306a40?})
	/usr/local/go/src/go/ast/walk.go:26 +0x91
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6bfd8?, 0xc00b8b3b40?})
	/usr/local/go/src/go/ast/walk.go:144 +0x36b
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6bf60?, 0xc0083526f0?})
	/usr/local/go/src/go/ast/walk.go:153 +0x17c5
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6be70?, 0xc008352720?})
	/usr/local/go/src/go/ast/walk.go:163 +0x1ca9
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6c1b8?, 0xc00b8b3b80?})
	/usr/local/go/src/go/ast/walk.go:74 +0x1b8d
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6c1e0?, 0xc008352750?})
	/usr/local/go/src/go/ast/walk.go:85 +0x218a
go/ast.Walk({0xd69ea0?, 0xc008d1eba0?}, {0xd6c618?, 0xc00b8ca330?})
	/usr/local/go/src/go/ast/walk.go:168 +0x16a5
sigs.k8s.io/controller-tools/pkg/loader.(*referenceSet).collectReferences(0xc00b8d34c0, {0xd6f430?, 0xc00b8ca330}, 0xc0043de050)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:107 +0xe5
sigs.k8s.io/controller-tools/pkg/loader.allReferencedPackages.func1(0xc00b093828?, 0x41005e1800?, 0xc00b8b3a40)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:164 +0x52
sigs.k8s.io/controller-tools/pkg/loader.(*typeVisitor).Visit(0x0?, {0xd6c690?, 0xc00b8b3a40?})
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/visit.go:64 +0xa7
go/ast.Walk({0xd69ec0?, 0xc00b736900?}, {0xd6c690?, 0xc00b8b3a40?})
	/usr/local/go/src/go/ast/walk.go:51 +0x62
go/ast.Walk({0xd69ec0?, 0xc00b736900?}, {0xd6c2d0?, 0xc00b8b3bc0?})
	/usr/local/go/src/go/ast/walk.go:344 +0x233c
go/ast.walkDeclList({0xd69ec0, 0xc00b736900}, {0xc00932b800?, 0xd, 0x0?})
	/usr/local/go/src/go/ast/walk.go:38 +0x91
go/ast.Walk({0xd69ec0?, 0xc00b736900?}, {0xd6c208?, 0xc004303d80?})
	/usr/local/go/src/go/ast/walk.go:366 +0xdea
sigs.k8s.io/controller-tools/pkg/loader.EachType(0xc00240f0a0, 0xc00b7368e8)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/visit.go:38 +0x96
sigs.k8s.io/controller-tools/pkg/loader.allReferencedPackages(0xc00240f0a0, 0xc0043de050)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:162 +0x275
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check(0xc001c563c0, 0xc00240f0a0)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:249 +0x187
sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1(0x63?)
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:257 +0x65
created by sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check
	/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/loader/refs.go:255 +0x1cd
make: *** [Makefile:47: generate] Error 2
```
I tried both on master and v0.17.0.

Any advice?

**Comment by jianhang33** [comment]
In webhook.go at line 154 (v0.17.0), there is a logging statement
```
		// An error here indicates bad configuration or an outage. Log for debugging.
		klog.Errorf("Failed to make webhook authenticator request: %v", err)
		return nil, false, err
```
Is this error message logged by default when authentication fails? A similar logging statement is needed at line 175 of webhook.go.

**Comment by dimityrmirchev** [comment]
> Adding log message to the code is something I can do, but when I tried to compile the code by running 'make all', I got some errors.

The `make all` command seem to be broken for me too, I should check that. Meanwhile you can use `make docker-build` to build the docker image or something similar to `docker build -t oidc-webhook-authenticator:my-tag .` if you want to tag the image with a custom tag.

> In webhook.go at line 154 (v0.17.0), there is a logging statement

I do not find this code in the project. Is this from the `kube-apiserver` codebase? If I follow the comment then this is not always logged when auth fails but when the requests for some reason do not reach the webhook authenticator which I think is not what is happening in your case since you are seeing `oidc-webhook-authenticator: "/validate-token" satisfied by exact match` logs.

**Comment by dimityrmirchev** [comment]
Closing since there is no longer activity on this issue.

**Comment by jianhang33** [comment]
I got my issue resolved by adding logging statements. Is there a reason why you don't want to put logging here? After I found out what was wrong my issue was easy to resolve.

**Comment by dimityrmirchev** [comment]
Hi @jianhang33 and thanks for the feedback! I am glad that you were able to solve your issue. W.r.t. the logs I think that it would be meaningful if we have some debugging logs that can be turned on/off. I will open an issue regarding this. Feel free to even open a PR if that is of interest to you, otherwise I will pick it up when I have some spare time.

**Comment by jianhang33** [comment]
@dimityrmirchev I am more than happy to create a PR, but I found that I don't know

1. how to access the verbose level set when starting the server
2. how to enable the logger for the level

If you have examples that I can follow, I'll make the change. Otherwise, thanks for taking care of it.

**Comment by dimityrmirchev** [comment]
@jianhang33 It seems that there was a bug when applying the logging configuration. This is probably why you did not see any logs when increasing the verbosity level. I opened https://github.com/gardener/oidc-webhook-authenticator/pull/111 which should fix this issue and add the additional logs. Mind that these logs can be enabled by setting `--zap-log-level` flag >= 10 when starting the server.

**Comment by jianhang33** [comment]
Thanks. I'll try it out when it is available, I assume, in tag v0.22.0.
